### ISSUE TITLE
Replace Bit::Vector

### ISSUE THREAD

**Issue by rgov** [stp/stp#4]
`Bit::Vector` aka `CONSTANTBV` is [LGPL-2](http://www.gnu.org/licenses/old-licenses/lgpl-2.0.html) licensed, and the implementation is somewhat cumbersome to use.

Here are the operations that are currently used, most of which would need to be reimplemented:

```
CONSTANTBV::BitVector_add
CONSTANTBV::BitVector_Bit_Off
CONSTANTBV::BitVector_Bit_On
CONSTANTBV::BitVector_bit_test
CONSTANTBV::BitVector_Boot
CONSTANTBV::BitVector_Chunk_Store
CONSTANTBV::BitVector_Clone
CONSTANTBV::BitVector_Compare
CONSTANTBV::BitVector_Concat
CONSTANTBV::BitVector_Copy
CONSTANTBV::BitVector_Create
CONSTANTBV::BitVector_decrement
CONSTANTBV::BitVector_Destroy
CONSTANTBV::BitVector_Dispose
CONSTANTBV::BitVector_Divide
CONSTANTBV::BitVector_Div_Pos
CONSTANTBV::BitVector_Empty
CONSTANTBV::BitVector_equal
CONSTANTBV::BitVector_Error
CONSTANTBV::BitVector_Fill
CONSTANTBV::BitVector_Flip
CONSTANTBV::BitVector_from_Bin
CONSTANTBV::BitVector_from_Dec
CONSTANTBV::BitVector_from_Hex
CONSTANTBV::BitVector_Hash
CONSTANTBV::BitVector_increment
CONSTANTBV::BitVector_Interval_Copy
CONSTANTBV::BitVector_is_empty
CONSTANTBV::BitVector_is_full
CONSTANTBV::BitVector_Lexicompare
CONSTANTBV::BitVector_Move_Left
CONSTANTBV::BitVector_Move_Right
CONSTANTBV::BitVector_msb_
CONSTANTBV::BitVector_Multiply
CONSTANTBV::BitVector_Mul_Pos
CONSTANTBV::BitVector_Negate
CONSTANTBV::BitVector_Resize
CONSTANTBV::BitVector_shift_right
CONSTANTBV::BitVector_Sign
CONSTANTBV::BitVector_sub
CONSTANTBV::BitVector_to_Bin
CONSTANTBV::BitVector_to_Dec
CONSTANTBV::BitVector_to_Hex
CONSTANTBV::Set_Complement
CONSTANTBV::Set_ExclusiveOr
CONSTANTBV::Set_Intersection
CONSTANTBV::Set_Max
CONSTANTBV::Set_Union
```

**Comment by msoos** [comment]
We have been discussing with @delcypher over this. The LGPL is kind of bad, especially since STP is advertised as MIT licensed. That library is pretty awful, but I don't feel like re-writing it. We'll keep this in mind, but for the moment, there seems to be no one with time and energy to do this, so it's a bit stale. Let me close this. If anyone feels like they have the time&energy to move forward with this, I'd be really happy of course!

**Comment by rgov** [comment]
Re-opening issue.

**Comment by rgov** [comment]
There's also [Boost::Multiprecision](http://www.boost.org/doc/libs/1_66_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints.html) which can be backed by GMP and other libraries.

**Comment by rgov** [comment]
@jzmaddock: Could we get your opinion? We have code that uses fixed-width 2's complement bit vectors, where the width is determined at runtime but doesn't change over the lifetime of an object. Can we use `cpp_int` in this way?

**Comment by msoos** [comment]
GMP is really not an option... that CONSTANTBV does some really wacky optimisations -- it's actually really clever. I have used GMP before, it's not the same thing, and I don't think it can be adapted in any way. I think we could re-write it in C++, essentially taking the rules that CONSTANTBV uses and adapting them to C++. Then we could add our own rules, etc. which would be really nice actually.

**Comment by jzmaddock** [comment]
>Could we get your opinion? We have code that uses fixed-width 2's complement bit vectors, where the width is determined at runtime but doesn't change over the lifetime of an object. Can we use cpp_int in this way?

cpp_int is sign-magnitude representation, but if you're performing all unsigned-integer bit-fiddling operations and don't reply on signed-integer wrap-around (which is undefined behaviour for regular integers), then you should be fine.  Actually, even if you did want to rely on wrap-around at N-bits, you could always just mask off the lower N bits after a potentially overflowing operation.  What cpp_int doesn't have is a concept of "max size set at runtime" - you can have either compile time fixed size, or arbitrary runtime precision - you could build runtime fixed precision on top of arbitrary precision though, especially as you will need to adapt the interface to you're current needs?

**Comment by rgov** [comment]
@msoos I'm not really sure I see Bit::Vector's"wacky optimizations," it looks to my (inexpert) eyes like a pretty straightforward implementation. TTMath has nice hand rolled assembly implementations but the author suggested I use Boost also.

"Just mask off the lower N bits after a potentially overflowing operation" seems inefficient, though.

**Comment by msoos** [comment]
Ah, I just looked through the code, you are right. Somehow I remembered this code to be very different, even performing a different function -- optimising away things like:

0 + N = N

so the output doesn't depend on N, in a symbolic way. I think I was very off. I'm sorry. Thanks for picking this up! By the way, I have been working a _lot_ on CMS lately, as I now work at a university, and I am tasked to improve it for certain scenarios. I might even get tasked to work on STP. So... hopefully, we'll see some interesting improvements. Thanks again for picking this up, and sorry for my confusing response!

**Comment by TrevorHansen** [comment]
So that I understand better, is the LGPL-2 a problem for us? Because we've not made any changes to the library, we or others can still release binaries of STP without limitation?

**Comment by andrewvaughanj** [comment]
I took a lightening-speed look at this, so my apologies if I've overlooked something:

- [TTMath](https://www.ttmath.org) -- BSD, compile-time width ❌ 

- [libtommath](https://github.com/libtom/libtommath) -- public domain, run-time width ✅ 

- [ARPREC](https://www.davidhbailey.com/dhbsoftware/) -- not free, didn't investigate about run-time width ❌  

- [GMP](https://gmplib.org) -- GPL, run-time width ❌ 

- [Boost](https://www.boost.org/doc/libs/1_66_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints.html) -- depends on back-end, `cpp_int` is compile-time width (as is `tom_int`, it seems if used via Boost) ❌  

- [OpenSSL BN](https://www.openssl.org/docs/man1.0.2/man3/bn.html) -- Apache, run-time width ✅ 

Given this, I guess that OpenSSL BN and libtommath are the front-runners (both have amenable licenses and both support run-time width). Importantly, they both also have a way to get an arbitrary bit inside of a run-time specified width term.

I'll take a look at what's in `Simplifier/constantBitP` and see if one solution seems more appropriate than the other. What I'd really like to do is create an abstraction over both, and then be able to change at runtime, but maybe just getting one working is good enough for now.

@rgov you mentioned OpenSSL BN on https://github.com/stp/stp/pull/358 -- do you have any thoughts on the appropriateness of BN to replace Bit::Vector?

**Comment by rgov** [comment]
I haven't used libcrypto (from OpenSSL) in a long time, so I couldn't say how it compares to the others. I would guess that others libraries take more liberties with acceleration because they don't have timing problems to worry about?

**Comment by TrevorHansen** [comment]
First some background. We're using Steffen Beyer's bit-vector library, called Bit::Vector. Currently we use version 6.4 from Sep-2006. You can download it and later versions from here: http://guest.engelschall.com/~sb/download/

The licence from the files we use from Bit::Vector is:

>  This library is free software; you can redistribute it and/or          */
> /*    modify it under the terms of the GNU Library General Public            */
> /*    License as published by the Free Software Foundation; either           */
> /*    version 2 of the License, or (at your option) any later version.       */
> 

The licence of the whole Bit:Vector package is:

> This package with all its parts is
> 
> Copyright (c) 1995 - 2004 by Steffen Beyer.
> All rights reserved.
> 
> This package is free software; you can use, modify and redistribute
> it under the same terms as Perl itself, i.e., under the terms of
> the "Artistic License" or the "GNU General Public License".
> 
> The C library at the core of this Perl module can additionally
> be used, modified and redistributed under the terms of the
> "GNU Library General Public License".
> 
> Please refer to the files "Artistic.txt", "GNU_GPL.txt" and
> "GNU_LGPL.txt" in this distribution, respectively, for details!

So by my reading, even though the files talk just about the LGPL, we can choose between any of the three licences.

The artistic licence seems reasonably permissive to me. 

Personally, I'm not sure what problems the licencing might cause people? If someone can explain, I'm happy to write to Steffen and ask what he wants us to do?

**Comment by rgov** [comment]
It is a bit of a confusing license, I think he intends that you could use the *perl* package under the standard perl package license, but if you want to use the C library (as we do), it is under LGPL.

In my reading of the LGPL, as a non-lawyer, it seems that you either need to use dynamic linking, or if using static linking (as we are), make it possible to link in new LGPL parts. Which is fine for the STP project itself, because anyone can grab the sources and rebuild and re-link it.

However, if someone statically links STP into their code, they must then also follow the LGPL requirement to allow Bit::Vector to be re-linked. 

In the commercial software development world, software licenses are a big headache, and many companies are forbidden from using GPL, especially GPLv3. LGPL is more friendly to them, but they need to know the license clearly before integrating with a product.

**Comment by TrevorHansen** [comment]
@rgov, when I read it, what sticks out to me is  the "additionally" part:

> The C library at the core of this Perl module can additionally
> be used, modified and redistributed under the terms of the "GNU Library General Public License".

If Bit::Vector could be used under the artistic licence, do you know if it makes everything OK?

**Comment by rgov** [comment]
It seems like the Artistic License 2.0 would be possibly OK but there are some red flags for the original license for being to vague (that have actually been tested in court, yikes).

**Comment by TrevorHansen** [comment]
Personally if Bit::Vector is licensed under either artistic licence, it seems enough to me. Both have clauses (see below) which mean people can modify and distribute STP pretty much however they want. So I'm happy with either artistic licence.

It sounds like, @rgov, that if Steffen says the C-part is licensed under the artistic licence 2, you'll be OK?

Does anyone else have a prefernce?


From 2.0:

> (8) You are permitted to link Modified and Standard Versions with other works, to embed the Package in a larger work of your own, or to build stand-alone binary or bytecode versions of applications that include the Package, and Distribute the result without restriction, provided the result does not expose a direct interface to the Package.

From 1.0:
> 8. Aggregation of this Package with a commercial distribution is always permitted provided that the use of this Package is embedded; that is, when no overt attempt is made to make this Package's interfaces visible to the end user of the commercial distribution. Such use shall not be construed as a distribution of this Package.

**Comment by rgov** [comment]
From reading Wikipedia, it sounds like there were problems with the original version that led to the Open Source Initiative rejecting it. The revised version addresses those problems. I'd rather use 2.0 if we're not going to replace it. We need to include the license in the STP license file.

**Comment by TrevorHansen** [comment]
Because the licence that Bit::Vector includes in its package is the
artistic licence 1.0. It seems that we have just one question for Steffen:
Is the C code in Bit::Vector licensed under the artistic licence 2.0?

**Comment by andrewvaughanj** [comment]
Okay, so:

1. I emailed Steffan to ask for his interpretation of "additionally be used" (my personal interpretation is that he is saying that the C portions can be licensed under *any* of three)

2. For sure, what's in STP today is "bad" because it is explicitly LGPL, so, at the very least, we should look at migrating the current version of `extlib-constbv` to use Bit-Vector-7.4 (which at least has the "ambiguous" three licenses)

3. We need to decide if we're happy having portions of STP licensed under the [Artistic License](https://dev.perl.org/licenses/artistic.html) -- comments further below

4. Even the most recent version of Bit-Vector *is not* under the Artistic License 2.0, but under 1.0 -- the statement about "and Distribute the result without restriction, provided the result does not expose a direct interface to the Package" only appears in 2.0, and is not in the license that Steffen includes in the most recent release (7.4)

I have an opinion on 3 that saying STP is using Bit::Vector under the terms of the Artistic License makes STP's overall claim of being MIT-licensed _still_ questionable. Clause 4a is _fine_, but isn't MIT-compatible (i.e., you need to do more under the Artistic License than you need to do under MIT). This makes the situation confusing for our users -- anyone who wants to use STP in their project (either dynamically or statically) needs to include both the MIT statement (for STP) and the Artistic License for Bit::Vector.

As a commercial user of STP, we ship STP as LGPL (because it statically includes LGPL components, therefore the whole STP binary is LGPL), and I would have no problem in shipping it STP as MIT + Bit::Vector components as Artistic License (and including the Artistic License text in our third-party license section). But, I do totally agree, it is a very "muddy" situation.

**Comment by andrewvaughanj** [comment]
Reply from Steffan:

> Dear Andrew:
> 
> The C library at the core of this Perl module can additionally be used,
> modified and redistributed under the terms of the "GNU Library General
> Public License".
> 
> The key word here is "additionally", which together with the word "can"
> is meant to mean "at your discretion".
> 
> This is because the GNU Library General Public License is somewhat less
> restrictive, and takes into account the fact that libraries are
> sometimes linked with software which may not be open source.
> 
> Otherwise, under the GNU General Public License, the software the
> library is linked with also has to be open source (GNU licensed),
> because of the "viral nature" of the GNU General Public License.
> 
> The GNU Library General Public License takes this viral nature away.
> 
> Summing up, you can freely choose between Artistic + GNU General Public
> License and/or the GNU Library General Public License.
> 
> The Artistic License was mostly adopted to ensure compatibility with
> Perl, but also because it is a fine license in itself IMHO.
> 
> Best regards, Steffen Beyer
>

**Comment by andrewvaughanj** [comment]
Follow-up: I replied to Steffen checking that he wants the code under Artistic 1.0 and not 2.0.

If there's a chance he wants it under 2.0, then the statement about "[Distribute the result without restriction, provided the result does not expose a direct interface to the Package](https://opensource.org/licenses/Artistic-2.0)" allows us to include it and remain MIT.

Honestly, that clause pretty much makes the library "free" for a lot of uses (you don't even need to _credit_ that you're using it, based on that one clause).

**Comment by andrewvaughanj** [comment]
Reply from Steffen:

> Dear Andrew,
> 
> I wasn't aware of the Artistic License 2.0, because I have stopped coding
> many years ago.
> 
> I really don't care that much about the specifics of the various licenses,
> as long as their basis consists of a fair amount of fair use, of give and
> take.
> 
> If I understood correctly, the AL 2.0 is similar to the GNU Library GPL in
> that it allows linking with non-open source software.
> 
> If so, I have nothing against someone who does the maintenance of my
> modules to add "the Artistic License (version 1.0 or 2.0, at your option)"
> to the current licensing conditions.
> 
> Or do it yourself, you can cite me.
> 
> Met vriendelijke groet,
> 
> Steffen Beyer

So maybe the AL 2.0 is "okay", but it still means that we cannot claim STP is regular MIT, because it contains this "weird mix" for AL 2.0.

I thought about asking Steffen if he would consider licensing Bit::Vector under MIT, but I don't think that meets his requirement of "basis consists of a fair amount of fair use, of give and take" (because MIT does allow you to "take" without "giving").

The binary of the clause of AL 2.0 seemed reasonable to me, but then I realised:

- we're not _just_ shipping it as binary (because we're distributing the STP code, which includes it)

- we _also_ have our own modifications (e.g., putting it in a namespace, changes for `THREAD_LOCAL`, etc.)

Basically, even if we to a version of Bit::Vector that is AL 2.0, we still need to resolve how to tell users of STP what license they need to adhere to (which might be MIT + AL 2.0).

**Comment by rgov** [comment]
First off, the 2.0 license says:

> (8) You are permitted to link Modified and Standard Versions with other works, to embed the Package in a larger work of your own, or to build stand-alone binary or bytecode versions of applications that include the Package, and Distribute the result without restriction, provided the result does not expose a direct interface to the Package.

That seems to give us a lot of room here to do what we want, since we are "embedding the Package into a larger work" (STP) and users might be "building stand-alone binary ... versions of applications that include the Package".

To go a little further, since @andrewvaughanj says we have some code changes in our copy, then we have a "Modified Version" according to the license.

> (6) You may Distribute a Modified Version in Compiled form without the Source, provided that you comply with Section 4 with respect to the Source of the Modified Version.

I think we just need to include in the License file:

> Uses a modified version of Bit::Vector by Steffen Beyer, originally released under the Artistic License 2.0 and 
> available at http://guest.engelschall.com/~sb/download/, with application programming interface changes.
>
> <<Artistic License 2.0 text>>

Then we need to do one of 4(a),(b),(c) which is easy, as we already do this:

> (a) make the Modified Version available to the Copyright Holder of the Standard Version, under the Original License, so that the Copyright Holder may include your modifications in the Standard Version.

I believe that is sufficient to say that:

* Users are allowed to link in STP without needing to provide the source code to Bit::Vector or how to get it.
* Users are allowed to distribute the source of STP, as the License file (and file header) explain the license.

Then I would be fine with saying that STP is just "MIT" in most places except the License file.

**Comment by andrewvaughanj** [comment]
Yep; I think that's actually quite good.

- Source distributions need to maintain the notice about including AL 2.0 components, and that's addressed with what you wrote

- Binary distributions (e.g., as a user of STP might distribute) then fall under the Clause 8, which has no obligations (apart from what's currently required for MIT for STP)

Do we think we need to update Bit::Vector though? Steffan's email seems to suggest that he's happy for any version of Bit::Vector to retroactively be re-licensed as AL 2.0.

**Comment by rgov** [comment]
We should probably look at what changes there are and see if we want to bring them in. Of course we will need to make sure we have some modifications so we fall under the Modified Version exception :)

**Comment by andrewvaughanj** [comment]
Okay, I can take a look at this.

@msoos what are your thoughts here? You had a quite a strong opinion in 2018, so I'd be interested to know if you've been "won around" by what's been proposed.

**Comment by TrevorHansen** [comment]
I'm happy to keep using Steffen Beyer's bit-vector library. The licensing seems OK to me. If anyone wants to discuss further, feel free to reopen this.

**Comment by aytey** [comment]
I'm re-opening this; we (the STP maintainers) have correspondence from Steffen that's he okay for us to:

- officially relicense Bit::Vector on our GitHub organisation under MIT

What we then need to do is:

- rip-out the copies of Bit::Vector from STP, build Bit::Vector as an external dependency, link this in

I haven't re-read this thread to remind myself, but I think AL is incompatible with MIT, so currently we cannot call STP "MIT licensed" because of AL infecting it (you're required to do more under AL than MIT, so anyone who uses STP and thinks they can follow MIT is incorrect).
